modules/gapi/src/executor/thread_pool.hpp

```diff
@@ -0,0 +1,70 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2024 Intel Corporation
+
+#ifndef OPENCV_GAPI_THREAD_POOL_HPP
+#define OPENCV_GAPI_THREAD_POOL_HPP
+
+#include <functional>
+#include <vector>
+#include <thread>
+#include <mutex>
+#include <atomic>
+#include <condition_variable>
+
+#include <opencv2/gapi/own/exports.hpp> // GAPI_EXPORTS
+
+#if defined(HAVE_TBB)
+#  include <tbb/concurrent_queue.h> // FIXME: drop it from here!
+template<typename T> using QueueClass = tbb::concurrent_bounded_queue<T>;
+#else
+#  include "executor/conc_queue.hpp"
+template<typename T> using QueueClass = cv::gapi::own::concurrent_bounded_queue<T>;
+#endif // TBB
+
+namespace cv {
+namespace gapi {
+namespace own {
+
+class WaitGroup {
+public:
+    void add();
+    void done();
+    void wait();
+
+private:
+    uint64_t                task_counter{0u};
+    std::mutex              m;
+    std::condition_variable all_done;
+};
```

Comment:
Is this what you've been looking for?

https://en.cppreference.com/w/cpp/thread/barrier

Reply:
> c++20

Reply:
No, the difference there is that `WaitGroup` is working with arbitrary number of `tasks` that can be added dynamically while the `Barrier` knows upfront how many tasks it waits.

Reply:
As we discussed, we know our graph structure before hand, so we know upfront how many tasks we should wait

The suggestion is to remove `wait()` and wait for the completion outside the `ThreadPool` (e.g by using something like barrier)